### ISSUE TITLE
Follows from closed span leads to panic

### ISSUE THREAD

**Issue by TmLev** [tokio-rs/tracing-opentelemetry#14]
### Version

```
tracing-opentelemetry-follows-from-closed-span-repro v0.1.0 (/.../tracing-opentelemetry-follows-from-closed-span-repro)
├── tracing v0.1.37
│   ├── tracing-attributes v0.1.23 (proc-macro)
│   └── tracing-core v0.1.30
├── tracing-opentelemetry v0.18.0
│   ├── tracing v0.1.37 (*)
│   ├── tracing-core v0.1.30 (*)
│   ├── tracing-log v0.1.3
│   │   └── tracing-core v0.1.30 (*)
│   └── tracing-subscriber v0.3.16
│       ├── tracing v0.1.37 (*)
│       ├── tracing-core v0.1.30 (*)
│       └── tracing-log v0.1.3 (*)
└── tracing-subscriber v0.3.16 (*)
```

### Platform

macOS 12.4, M1 Pro processor

### Description

Calling `tracing::Span::current().follows_from(closed_span_id)` with ID of an already closed span leads to panic:

```logs
thread 'main' panicked at 'Span to follow not found, this is a bug', /.../.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-opentelemetry-0.18.0/src/layer.rs:779:14
```

I created a small repro demonstrating the issue:
https://github.com/TmLev/tracing-opentelemetry-follows-from-closed-span-repro

I expected no such panics, because it seems to me that `follows_from` is perfectly suitable for instantiating "async" causal relations even if the original span is already closed.

Instead, `tracing_opentelemetry` panics.

I believe the relevant piece of code is this:

https://github.com/tokio-rs/tracing-opentelemetry/blob/13a2ad1065915a682881e17bce61571b75f65fde/src/layer.rs#L777-L790

**Comment by TmLev** [comment]
I'm not an expert in OpenTelemetry, so I can't offer a PR fixing this issue. Seems like we need to create a OpenTelemetry context and fill it appropriately (that's the part I'm struggling with), then pass it as `follows_context` to `otel::Link::new()`.

**Comment by jtescher** [comment]
@TmLev yeah you are correct, the issue is that the otel context for the closed span is no longer available so it can't look it up. You can add your link directly via [OpenTelemetrySpanExt::add_link](https://docs.rs/tracing-opentelemetry/latest/tracing_opentelemetry/trait.OpenTelemetrySpanExt.html#tymethod.add_link), the docs have some examples, but we should probably warn and ignore if the tracing span has been closed rather than causing a panic.

Are you open to submitting a PR to that effect?

**Comment by TmLev** [comment]
I'm not sure warning and ignoring is the right thing to do here. I'd expect to be able to make a causal relationship between closed span and some other one.

As for the PR, I can definitely try! But once again, I'm no expert in OpenTelemetry and it may take some time for me to figure out all the details.

**Comment by jtescher** [comment]
> I'm not sure warning and ignoring is the right thing to do here. I'd expect to be able to make a causal relation between closed span and some other one.

Feel free to look into other solutions as well, the current architecture stores the otel state in the tracing registry which means the context is cleaned up after the span ends. This requires you to either store the state yourself if you want to associated past span data with current spans, or otherwise update the storage of otel contexts.

**Comment by jnicholls** [comment]
I think this is a similar issue/root cause to a panic I'm experiencing, which is attempting to extend a `Span` into another scoped thread, i.e.

```
#[instrument]
fn do_stuff_blocking() {
    std::thread::scope(|s| {
        let current_span = Span::current();
        s.spawn(move || current_span.in_scope(|| {
            tokio::runtime::Builder::new_current_thread()
                .enable_all().build().expect("c'mon man!")
                .block_on(async {
                    // do some async stuff
                }.in_current_span())
        }))
        .join();
    });
}
```

This results in the tracer layer panicking, I presume because of a `Span` already being closed and thus the OpenTelemetry context is lost? But given this is a scoped thread, the outer (i.e. `Span::current()`) span should not be closed by that time. Thoughts? In the meantime I'll try to narrow down where the panic is occurring (this is currently running in a lambda via `cargo lambda watch` which I think is hiding the panic details).

**Comment by jnicholls** [comment]
Random thought: Could this be perhaps from the span crossing two different Tokio runtimes? I'll check into that right now.

**Comment by jtescher** [comment]
@jnicholls yeah span data is in the tracing subscriber registry so going across runtimes will not work well. We should probably remove the expects generally and have a noop or warn in those cases.

**Comment by TmLev** [comment]
Am I correct in understanding that #86 will cause `span.follows_from(closed_span_id)` calls to be ignored, resulting in OpenTelemetry being unaware of the spans' relationships?

**Comment by guswynn** [comment]
@TmLev yes! Spans can be disabled (`Span::id` even returns an `Option<Id>` and `Span::follows_from` takes an `Into<Option<Id>>`) and `follows_from` specifically documents this (https://docs.rs/tracing/latest/tracing/struct.Span.html#method.follows_from)

> If this span is disabled, or the resulting follows-from relationship would be invalid, this function will do nothing.

**Comment by TmLev** [comment]
I don't think this approach is appropriate for the OpenTelemetry layer. According to the specification (https://github.com/opentracing/specification/blob/master/specification.md#references-between-spans), the 'follows from' relationship does not require the 'parent' span to be available, open, or unclosed in order to establish a relationship.

**Comment by guswynn** [comment]
@TmLev The structure of the tracing/tracing-subscriber crates means that a successful `Span::follows_from` call requires a follows span that is both alive and can be seen by the layer (i.e. its not per-layer filtered away). My pr changes this crate from panicking/aborting, to skipping, which is more in line with the expected behavior of a tracing `Layer` implementation.

Relationships between parents/follows and children that can't be captured by this model, like some of the ones given in the spec link can be added with: https://docs.rs/tracing-opentelemetry/latest/tracing_opentelemetry/trait.OpenTelemetrySpanExt.html#tymethod.add_link . This is a perfect usecase for that extension crate, as you can obtain a `SpanContext` from a follows span (`follows.context().span().span_context().clone()`) and continue to use it using this extension trait after that follows span has closed. I consider this to be very similar to using `OpenTelemetrySpanExt` to link/parent spans across multiple processes (another relationship that can't be expressed using the `tracing` crate running in one process alone)

**Comment by TmLev** [comment]
> The structure of the tracing/tracing-subscriber crates means that a successful Span::follows_from call requires a follows span that is both alive and can be seen by the layer (i.e. its not per-layer filtered away).
> ...
> Relationships between parents/follows and children that can't be captured by this model

Would it be worthwhile to reconsider the current behaviour, partly to align it more closely with the OpenTelemetry approach to tracing? Or is such a change beyond the intended scope of the `tracing`/`tracing-subscriber` libraries?

**Comment by jtescher** [comment]
Resolved in #86
